core/components/tree/basenode.js

```diff
@@ -0,0 +1,1151 @@
+    }
+  } else {
+    ce = this.getChildrenElement();
+    if (ce) {
+      Blockly.utils.style.setElementShown(ce, false);
+    }
+  }
+  if (el) {
+    this.updateIcon_();
+  }
+
+  if (isStateChange) {
+    if (expanded) {
+      this.doNodeExpanded();
+    } else {
+      this.doNodeCollapsed();
+    }
+  }
+};
+
+/**
+ * Used to notify a node of that we have expanded it.
+ * Can be overidden by subclasses, see Blockly.tree.TreeNode.
+ */
+Blockly.tree.BaseNode.prototype.doNodeExpanded = function() {
+  // NOP
+};
+
+/**
+ * Used to notify a node that we have collapsed it.
+ * Can be overidden by subclasses, see Blockly.tree.TreeNode.
+ */
+Blockly.tree.BaseNode.prototype.doNodeCollapsed = function() {
+  // NOP
+};
+
+/**
+ * Toggles the expanded state of the node.
+ */
+Blockly.tree.BaseNode.prototype.toggle = function() {
+  this.setExpanded(!this.getExpanded());
+};
+
+/**
+ * Expands the node.
+ */
+Blockly.tree.BaseNode.prototype.expand = function() {
+  this.setExpanded(true);
+};
+/**
+ * Collapses the node.
+ */
+Blockly.tree.BaseNode.prototype.collapse = function() {
+  this.setExpanded(false);
+};
+
+/**
+ * Collapses the children of the node.
+ */
+Blockly.tree.BaseNode.prototype.collapseChildren = function() {
+  this.forEachChild(function(child) { child.collapseAll(); });
+};
+
+/**
+ * Collapses the children and the node.
+ */
+Blockly.tree.BaseNode.prototype.collapseAll = function() {
+  this.collapseChildren();
+  this.collapse();
+};
+
+/**
+ * Expands the children of the node.
+ */
+Blockly.tree.BaseNode.prototype.expandChildren = function() {
+  this.forEachChild(function(child) { child.expandAll(); });
+};
+
+/**
+ * Expands the children and the node.
+ */
+Blockly.tree.BaseNode.prototype.expandAll = function() {
+  this.expandChildren();
+  this.expand();
+};
+
+/**
+ * Expands the parent chain of this node so that it is visible.
+ */
+Blockly.tree.BaseNode.prototype.reveal = function() {
+  var parent = this.getParent();
+  if (parent) {
+    parent.setExpanded(true);
+    parent.reveal();
+  }
+};
+
+/**
+ * Sets whether the node will allow the user to collapse it.
+ * @param {boolean} isCollapsible Whether to allow node collapse.
+ */
+Blockly.tree.BaseNode.prototype.setIsUserCollapsible = function(isCollapsible) {
+  this.isUserCollapsible_ = isCollapsible;
+  if (!this.isUserCollapsible_) {
+    this.expand();
+  }
+  if (this.getElement()) {
+    this.updateExpandIcon();
+  }
+};
+
+/**
+ * @return {boolean} Whether the node is collapsible by user actions.
+ */
+Blockly.tree.BaseNode.prototype.isUserCollapsible = function() {
+  return this.isUserCollapsible_;
+};
+
+/**
+ * Creates HTML Element for the node.
+ * @return {!Element} html element
+ * @protected
+ */
+Blockly.tree.BaseNode.prototype.toDom = function() {
+  var childClass = this.config_.cssChildrenNoLines;
+
+  var nonEmptyAndExpanded = this.getExpanded() && this.hasChildren();
+
+  var children = document.createElement('div');
```

Comment:
This (and the previous file) are introducing some new uses of `document`.  Is that going to be an issue when running headless/on node?

Reply:
It shouldn't. We don't initialize a toolbox in headless mode, and all these methods are dom methods.